backend/benefit/applications/benefit_aggregation.py

```diff
@@ -0,0 +1,168 @@
+        return past_benefit_info
+
+    recent_benefits = _get_benefits_relevant_for_validation(
+        _get_past_benefits(application, company, social_security_number, end_date),
+        start_date,
+    )
+
+    past_benefit_info.validation_errors.extend(
+        _get_benefit_overlap_errors(recent_benefits, start_date, end_date)
+    )
+
+    past_benefit_info.months_used = sum(
+        benefit.duration_in_months for benefit in recent_benefits
+    )
+    if apprenticeship_program:
+        # Kanslia Helsinki-lisä Teams discussion 2021-12-09: there's no upper limit defined for
+        # sequentially granted apprenticeship benefits
+        past_benefit_info.months_remaining = None
+    else:
+        past_benefit_info.months_remaining = max(
+            Decimal(0), BENEFIT_MAX_DURATION - past_benefit_info.months_used
+        )
+
+    if (
+        not apprenticeship_program
+        and past_benefit_info.months_used + duration_in_months(start_date, end_date)
+        > Application.BENEFIT_MAX_MONTHS
+    ):
+        # granting this benefit would exceed the limit
+        past_benefit_info.validation_errors.append(
+            _("Benefit can not be granted before 24-month waiting period expires")
+        )
+    return past_benefit_info
+
+
+def _get_past_benefits(application, company, social_security_number, end_date):
+    """
+    Return a list containing two types of objects:
+    * PreviousBenefits
+    * Applications that are in ACCEPTED status
+
+    PreviousBenefit and Application objects have these have these in common:
+    * start_date and end_date fields
+    * duration_in_months property
+    """
+
+    # The waiting time starts at the end of the latest benefit period granted.
+    previously_accepted_applications = Application.objects.filter(
+        employee__social_security_number=social_security_number,
+        company=company,
+        status=ApplicationStatus.ACCEPTED,
+        start_date__lte=end_date,  # catch also overlapping benefits
+    )
+
+    if application:
+        previously_accepted_applications = previously_accepted_applications.exclude(
+            pk=application.pk
+        )
+
+    previous_benefits = PreviousBenefit.objects.filter(
+        social_security_number=social_security_number,
+        company=company,
+        start_date__lte=end_date,
+    )
+    return sorted(
+        chain(previously_accepted_applications, previous_benefits),
+        key=operator.attrgetter("start_date"),
+        reverse=True,
+    )  # most recent first
+
+
+def _get_benefits_relevant_for_validation(past_benefits, start_date):
+    """
+    :param past_benefits: list of PreviousBenefit and/or Application objects
+    :param start_date:
+    :return:
+    """
+    if not past_benefits:
+        return []
+    most_recent_benefit = past_benefits[0]
+    if (
+        not most_recent_benefit
+        or most_recent_benefit.end_date
+        < start_date - relativedelta(months=BENEFIT_WAITING_PERIOD_MONTHS)
```

Comment:
I have a hard time reading this without being confused as I read the line
`or most_recent_benefit.end_date`
on its own.

Perhaps adding parentheses or doing something else
stylistically might help to make this more readable,
or putting the `start_date - ...` part into a variable and
then using that in the comparison to make it fit on one
line might help.

I leave it up to you to decide what to do here, if anything.

Reply:
fixed

Reply:
There are more places with similar readability issues but I don't see them as crucial, I leave them up to you whether you want to change them or not. This is ok as it is now for me.